src/mozilla_sec_eia/utils.py

```diff
@@ -78,12 +97,126 @@ def getconn() -> pg8000.dbapi.Connection:
 
+    def _get_local_path(self, cache_directory: Path, filing: pd.Series) -> Path:
+        """Return path to a filing in local cache based on metadata."""
+        return cache_directory / Path(
+            f"{filing['CIK']}-{filing['year_quarter']}-"
+            f"{filing['Filename'].replace('edgar/data/', '').replace('/', '-')}"
+        )
+
+    def _v_stack_images(self, *images):
+        """Generate composite of all supplied images."""
+        # Get the widest width.
+        width = max(image.width for image in images)
+        # Add up all the heights.
+        height = sum(image.height for image in images)
+        composite = Image.new("RGB", (width, height))
+        # Paste each image below the one before it.
+        y = 0
+        for image in images:
+            composite.paste(image, (0, y))
+            y += image.height
+        return composite
+
+    def get_filing(
+        self,
+        filing: pd.Series,
+        cache_directory: Path = Path("./sec10k_filings"),
+    ) -> Path:
+        """Cache a single filing and return path."""
+        # Create cache directory
+        cache_directory.mkdir(parents=True, exist_ok=True)
+
+        blob = self.get_blob(filing["year_quarter"], filing["Filename"])
+        local_path = self._get_local_path(cache_directory, filing)
+
+        if exists := local_path.exists():
+            blob.update()
+            local_hash = _compute_md5(local_path)
+            remote_hash = blob.md5_hash
+            refresh = remote_hash != local_hash
+
+        if (not exists) or refresh:
+            blob.download_to_filename(local_path)
+
+        return local_path
+
+    def get_filings(
+        self,
+        filing_selection: pd.DataFrame,
+        cache_directory: Path = Path("./sec10k_filings"),
+    ) -> typing.Generator[Path, None, None]:
+        """Caches filings locally for quick access, and returns generator producing paths.
+
+        Args:
+            filing_selection: Pandas dataframe with same schema as metadata df where each row
+                is a filing to return.
+            cache_directory: Path to directory where filings should be cached.
+        """
+        for _, filing in filing_selection.iterrows():
+            yield self.get_filing(filing, cache_directory)
+
+    def get_ex_21_from_filing(
+        self,
+        filing: Path,
+    ) -> Image:
+        """Extract exhibit 21 from raw filing and return PIL Image of table."""
+        ex_21_pat = re.compile(
+            r"<DOCUMENT>\s?<TYPE>EX-21(\.\d)?([\S\s])*?(</DOCUMENT>)"
+        )
+        with filing.open() as f:
+            ex_21_text = ex_21_pat.search(f.read()).group(0)
+
+        pdf_bytes = io.BytesIO()
+        res = pisa.CreatePDF(ex_21_text, pdf_bytes)
+        if res.err:
+            raise RuntimeError(
+                f"Failed to create PDF from filing {str(filing)} exhibit 21."
+            )
+
+        doc = fitz.open(stream=pdf_bytes, filetype="pdf")
+        pages = [
```

Comment:
Can you add an error message if an image can't be generated? I'm experimenting with the file `edgar/data/1177274/0001079973-10-000019.txt` (CIK = 1177274, Rodobo International Inc, 2010 Q1) and in that filing the Ex. 21 table is missing even though there's an Ex. 21 attachment (it seems like the whole second half of the filing is crossed out). When I try to generate the Ex. 21 image for that filing I get the error:

```
Could not get image data from src attribute: graph1v2.jpg
'<img src="graph1v2.jpg" alt=""/>'
```

It would be nice if the error message said which filing couldn't generate an Ex. 21 image and then continued trying with the next filing in the filing selection.

Reply:
Changed to log a warning